src/lib.rs

```diff
@@ -1,127 +1,113 @@
 
-// We put a new-type around serde_cbor::Error in `mod decode` and `mod encode`,
-// because this keeps us compatible with how rmp-serde was exporting its errors,
-// and avoids unnecessary code changes.
+pub use prost::{self, DecodeError, EncodeError, Message};
+
+/// Decoding-specific types, here for backwards compatibility.
 pub mod decode {
+    use super::*;
+    use alloc::{
+        format,
+        string::{String, ToString},
+    };
+    use ciborium::de::Error as CiboriumError;
+    use core::fmt::Debug;
+
+    /// An error structure for CBOR decoding
     #[derive(Debug)]
-    pub struct Error(serde_cbor::Error);
+    pub struct Error(String);
 
-    impl core::fmt::Display for Error {
-        fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
-            write!(f, "Cbor Decode Error: {}", self.0)
+    impl Display for Error {
+        fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+            write!(f, "CBOR Decode Error: {:#?}", self.0)
         }
     }
 
-    impl From<serde_cbor::Error> for Error {
-        fn from(src: serde_cbor::Error) -> Self {
-            Self(src)
+    impl<T: Debug> From<CiboriumError<T>> for Error {
+        fn from(src: CiboriumError<T>) -> Self {
+            let err_msg = match src {
+                CiboriumError::Io(underlying) => format!("IO Error: {underlying:#?}"),
+                CiboriumError::Syntax(offset) => format!("Syntax Error at byte {offset:#?}"),
+                CiboriumError::Semantic(offset, inner_msg) => {
+                    format!("Semantic Error at byte {offset:#?}: {inner_msg:#?}")
+                }
+                CiboriumError::RecursionLimitExceeded => "Recursion limit exceeded".to_string(),
+            };
+
+            Self(err_msg)
         }
     }
 }
 
+/// Encoding-specific types, here for backwards compatibility.
 pub mod encode {
-    #[derive(Debug)]
-    pub struct Error(serde_cbor::Error);
+    use super::*;
 
-    impl core::fmt::Display for Error {
-        fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
-            write!(f, "Cbor Encode Error: {}", self.0)
-        }
-    }
+    /// CBOR encoding errors (this shouldn't actually be seen in practice)
+    #[derive(Debug)]
+    pub struct Error;
 
-    impl From<serde_cbor::Error> for Error {
-        fn from(src: serde_cbor::Error) -> Self {
-            Self(src)
+    impl Display for Error {
+        fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+            write!(f, "CBOR Encode Error")
         }
     }
 }
 
 /// Serialize the given data structure.
 ///
-/// Forward mc_util_serial::serialize to bincode::serialize(..., Infinite)
-/// Serialization can fail if `T`'s implementation of `Serialize` decides to
-/// fail.
+/// This method produces CBOR-encoded bytes.
 pub fn serialize<T: ?Sized>(value: &T) -> Result<Vec<u8>, encode::Error>
 where
     T: Serialize + Sized,
 {
-    Ok(serde_cbor::to_vec(value)?)
+    let mut writer = Vec::new();
```

Comment:
i wonder if we can preallocate this vec? i remember protobuf had a way to estimate

Reply:
I dunno, `2 * sizeof::<T>()` is maybe a decent heuristic...

Reply:
yeah i dont see an obvious way to do it from the lib documentation